### DISCUSSION TITLE
How to fix versioning for docker images with multiple compatibility parts

### DISCUSSION THREAD

**Discussion by mvitz** [renovatebot/renovate#30366] [Request Help] [answered]
### What would you like help with?

Other

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab, latest version

### Please tell us more about your question or problem

I use the `bellsoft/liberica-runtime-container` in my Java-based projects.

Unfortunately, they decided to use a somewhat special versioning for their containers:
> [JDK type]-Java version-[crac]-[cds]-[slim]-[OS release]-[libc type]-[date]
(see https://hub.docker.com/r/bellsoft/liberica-runtime-container within the *Tags*-Section).

For that reason, the default versioning of renovate does not work, and I tried to specify my own `packageRules` for that:

```
  "packageRules": [
    {
      "matchDatasources": [
        "docker"
      ],
      "matchPackageNames": [
        "bellsoft/liberica-runtime-container"
      ],
      "versioning": "regex:^(jdk|jre)-(?<major>\\d+)?(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?([\\._+](?<build>(\\d\\.?)+))?(-(?<compatibility>.*))?$"
    }
  ]
```

This does work correct because the leading _JDK type_ is constantly changed to the one that was pushed into the repository last.

Is there any way I can achieve that kind of versioning using regex or any other way?
Or would that require a change/addition within renovate (e.g. allowing multiple compatibility parts)?

**Comment by rarkins** [comment]
Can you give some examples of:
- current tag
- some valid updates (you would want Renovate to raise)
- some updates you'd want Renovate to skip, and why

**Reply by mvitz** [reply]
Yes sure:

Current tags:
 * `jre-21.0.3_11-cds-slim-stream-musl`
 * `jdk-21.0.3_11-stream-musl`

Valid updates for the given tags:
 * `jre-21.0.4_9-cds-slim-stream-musl`
 * `jdk-21.0.4_9-stream-musl`

some updates you'd want Renovate to skip, and why
 * `jdk-22.0.2_11-stream-musl` because JDK 22 is not a LTS release.

The main problem here is that within the tag the `jdk`- or `jre`-Prefix is important and should be preserved, as should the suffixes (like `-cds-slim-stream-musl` or `-stream-musl`).

**Reply by rarkins** [reply]
The "double" compatibility strings (both a prefix and suffix) make this hard to satisfy with only one rule.

If you were to use two rules then it would be something like this:

```json
{
  "matchCurrentValue": "/^jre-/",
  "versionCompatibility": "^jre-(?<version>.*)-(?<compatibility>.*)$",
  "versioning": "loose"
},
{
  "matchCurrentValue": "/^jdk-/",
  "versionCompatibility": "^jdk-(?<version>.*)-(?<compatibility>.*)$",
  "versioning": "loose"
}
```

The next issue is that we don't have first-class java LTS awareness so you'd need to constraint that manually for now. The final rules might look like this:

```json
{
  "matchCurrentValue": "/^(jdk|jre)-/",
  "versioning": "loose",
  "allowedVersions": "/^21\\."
},
{
  "matchCurrentValue": "/^jre-/",
  "versionCompatibility": "^jre-(?<version>.*)-(?<compatibility>.*)$"
},
{
  "matchCurrentValue": "/^jdk-/",
  "versionCompatibility": "^jdk-(?<version>.*)-(?<compatibility>.*)$"
}
```

You also may need to add a matchPackageNames to the last two in case you want the match to be more exact

**Reply by mvitz** [reply, marked as answer]
Thanks a lot. I added
```
    {
      "description": "Enable JDK updates of bellsoft/liberica-runtime-container.",
      "matchCurrentValue": "/^jdk-/",
      "matchDatasources": [
        "docker"
      ],
      "matchPackageNames": [
        "bellsoft/liberica-runtime-container"
      ],
      "versioning": "regex:^jdk-(?<major>\\d+)?(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?([\\._+](?<build>(\\d\\.?)+))?(-(?<compatibility>.*))?$"
    },
    {
      "description": "Enable JRE updates of bellsoft/liberica-runtime-container.",
      "matchCurrentValue": "/^jre-/",
      "matchDatasources": [
        "docker"
      ],
      "matchPackageNames": [
        "bellsoft/liberica-runtime-container"
      ],
      "versioning": "regex:^jre-(?<major>\\d+)?(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?([\\._+](?<build>(\\d\\.?)+))?(-(?<compatibility>.*))?$"
    },
    {
      "allowedVersions": "/^(jdk|jre)-(?:8|11|17|21)(?:\\.|-|$)/",
      "description": "Limit bellsoft/liberica-runtime-container versions to LTS releases.",
      "matchDatasources": [
        "docker"
      ],
      "matchPackageNames": [
        "bellsoft/liberica-runtime-container"
      ]
    }
```
to our internal java preset and this works as expected.

What are your thoughts about extending the included `workarounds:javaLTSVersions` preset?

**Reply by rarkins** [reply]
Sure, if you think it's a common case for others!

**Reply by HonkingGoose** [reply]
[DockerHub for `bellsoft/liberica-runtime-container` image](https://hub.docker.com/r/bellsoft/liberica-runtime-container) says:

> Pulls 500K+

I'll let you decide if 500K+ pulls means you want the workaround in the preset. 😉

**Reply by rarkins** [reply]
Yes, that sounds like enough

**Reply by HonkingGoose** [reply]
See issue:

- #31071

@mvitz if you want to be assigned to the issue, please comment in the _issue_. If not, I'll leave the issue open as a good first issue for newcomers to the project. 😉
